primitives/consensus/aura/src/block_proposal.rs

```diff
@@ -0,0 +1,73 @@
+use crate::InherentDigest;
+use futures::FutureExt;
+use sp_consensus::{Environment, Proposer};
+use sp_inherents::InherentData;
+use sp_runtime::traits::Block as BlockT;
+use sp_runtime::{Digest, DigestItem};
+use std::future::Future;
+use std::marker::PhantomData;
+use std::time;
+
+/// Proposer factory for PartnerChainsProposer. Allows passing ID: InherentDigest type parameter.
+pub struct PartnerChainsProposerFactory<B: BlockT, E: Environment<B>, ID> {
+	env: E,
+	phantom_data: PhantomData<(B, ID)>,
+}
+
+impl<B: BlockT, E: Environment<B>, ID> PartnerChainsProposerFactory<B, E, ID> {
+	pub fn new(env: E) -> Self {
+		Self { env, phantom_data: PhantomData }
+	}
+}
+
+impl<B: BlockT, E: Environment<B>, ID: InherentDigest + Send + Sync + 'static> Environment<B>
+	for PartnerChainsProposerFactory<B, E, ID>
+{
+	type Proposer = PartnerChainsProposer<B, E::Proposer, ID>;
+	type CreateProposer =
+		Box<dyn Future<Output = Result<Self::Proposer, Self::Error>> + Send + Unpin + 'static>;
+	type Error = <E as Environment<B>>::Error;
+
+	fn init(&mut self, parent_header: &<B as BlockT>::Header) -> Self::CreateProposer {
+		Box::new(self.env.init(parent_header).map(|res| {
+			res.map(|proposer| PartnerChainsProposer::<B, E::Proposer, ID>::new(proposer))
+		}))
+	}
+}
+
+/// Wraps a Proposer. Adds inherent data digest to the original logs and calls wrapped Proposer.
+pub struct PartnerChainsProposer<B: BlockT, P: Proposer<B>, ID: InherentDigest> {
+	pub proposer: P,
+	phantom_data: PhantomData<(B, ID)>,
+}
+
+impl<B: BlockT, P: Proposer<B>, ID: InherentDigest> PartnerChainsProposer<B, P, ID> {
+	pub(crate) fn new(proposer: P) -> Self {
+		Self { proposer, phantom_data: PhantomData }
+	}
+}
+
+impl<B: BlockT, P: Proposer<B>, ID: InherentDigest> Proposer<B>
+	for PartnerChainsProposer<B, P, ID>
+{
+	type Error = <P as Proposer<B>>::Error;
+	type Proposal = <P as Proposer<B>>::Proposal;
+	type ProofRecording = <P as Proposer<B>>::ProofRecording;
+	type Proof = <P as Proposer<B>>::Proof;
+
+	fn propose(
+		self,
+		inherent_data: InherentData,
+		inherent_digests: Digest,
+		max_duration: time::Duration,
+		block_size_limit: Option<usize>,
+	) -> Self::Proposal {
+		let mut logs: Vec<DigestItem> = Vec::from(inherent_digests.logs());
+		// It is a programmatic error to try to propose a block that has inherent data from with declared InherentDigest cannot be created.
+		let mut inherent_logs = ID::from_inherent_data(&inherent_data)
+			.expect("InherentDigest can be created from inherent data");
```

Comment:
I was trying to return Self::Proposal here but I failed.
If the message in `expect` is not lying I think it's not bad to fail here, because it means that node was not wired properly.